### PR TITLE
Allow using `Slice` `Iterator`s in const Contexts

### PR DESCRIPTION
**Pull request by onestacked**
~~based on #102225 by @fee1-dead~~

### CODE REVIEW

File: library/core/src/slice/iter/macros.rs [hunk 1]
```diff
@@ -4,9 +4,11 @@
...
     // The way we encode the length of a ZST iterator, this works both for ZST
     // and non-ZST.
-    ($self: ident) => {
-        $self.ptr.as_ptr() as *const T == $self.end
-    };
+    ($self: ident) => {{
+        #![allow(unused_unsafe)] // we're sometimes used within an unsafe block
+        // SAFETY: FIXME
+        unsafe { ($self.ptr.as_ptr() as *const u8).offset_from($self.end as *const u8) == 0 }
+    }};
```
Comment (chriss0612) [inline library/core/src/slice/iter/macros.rs:11]: I will still need to make sure these are safe / Non UB.

If anyone has any Ideas for how to do it better please write a sugestion.

File: library/core/src/slice/iter/macros.rs [hunk 2]
```diff
@@ -17,22 +19,31 @@ macro_rules! len {
...
+        let end = $self.end;
         if size == 0 {
             // This _cannot_ use `unchecked_sub` because we depend on wrapping
             // to represent the length of long ZST slice iterators.
-            $self.end.addr().wrapping_sub(start.as_ptr().addr())
+            //end.addr().wrapping_sub(start.as_ptr().addr())
+            // SAFETY: todo / probably UB
+            unsafe {
+                ($self.end as *const u8).offset_from($self.ptr.as_ptr() as *const u8) as usize
+            }
```
Comment (chriss0612) [inline library/core/src/slice/iter/macros.rs:30]: Same as with is_empty above.

**Comment by the8472** [comment]
This is the 2nd PR blocked on that PR which in turn is blocked on a bunch of concerns one of which is blocked on another WIP PR. It could take many months for these things to percolate.

**Comment by chriss0612** [comment]
I know, but I made those changes mainly because I wanted to know how much would be required for them.
And since I already had working local version I thought I should do a (Draft) PR so that no one will do unnecessary work re-implementing this in the future.

**Comment by fee1-dead** [comment]
@chriss0612, you could r? @ghost when opening a PR so that the PR will not be automatically assigned a reviewer.

**Comment by chriss0612** [comment]
Currently includes #104100, but DoubleEndedIterator could be made #[const_trait] here and the one for loop that uses a range could be manually converted to a while.

For Slices of zero sized types this is currently blocked on #92512, since there is currently not way to find the calculate the length of a zero sized slice without UB. 

Theoretically `const_eval_select`could be used to panic for slices of ZST's at CTFE, but I don't think that is a good approach. 

So I think that #92512 should be figured out before this is considerd for merging.

File: library/core/src/slice/iter/macros.rs (hunk 1 above)
Comment (chriss0612) [reply library/core/src/slice/iter/macros.rs]: Currently not possible without UB, see #92512

**Comment by Dylan-DPC** [comment]
Closing this as it is an experiment and blocked on multiple things

**Pull request closed without merging**